include/xgboost/data.h

```diff
@@ -287,10 +289,23 @@ class SortedCSCPage : public SparsePage {
   explicit SortedCSCPage(SparsePage page) : SparsePage(std::move(page)) {}
 };
 
+class EllpackPageImpl;
+class EllpackPage {
```

Comment:
should this be more generalized so we can support multiple binned matrix types fairly easily. perhaps, make this an adapter, with a factory method to create the the underlying implementation:

```
struct binned_page {
   // common interface applicable to all pages
   virtual int num_bins() = 0;
   virtual int get_feature_bin(int bidx) = 0;
   virtual float get_feature_value(int fbin) = 0;
   // specifics for a page
   virtual float get_feature_value(ridx, fidx);
   // etc...
};

struct ellpack_page : binned_page {
   // realize interface
};

struct csr_page : binned_page {
   // realize interface
};

// etc...
static binned_page *create_binned_page(pertinent_data_after_quantile_generation);
```

Reply:
We can probably wait until we have another binned page.

Reply:
Agree, let's keep this in mind but not for this PR.